pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,234 @@
+		}
+
+		for _, endpoint := range toAddEndpoints {
+			if endpoint.IsCluster() {
+				if len(ps.clusterChecks) > 0 {
+					ps.startNewWorker(endpoint, ps.clusterChecks)
+				} else {
+					level.Debug(ps.logger).Log("msg", fmt.Sprintf("Skipped probing on %s: no cluster checks defined", endpoint.GetName()))
+				}
+			} else {
+				if len(ps.nodeChecks) > 0 {
+					ps.startNewWorker(endpoint, ps.nodeChecks)
+				} else {
+					level.Debug(ps.logger).Log("msg", fmt.Sprintf("Skipped probing on %s: no node checks defined", endpoint.GetName()))
+				}
+			}
+		}
+		ps.currentTopology = newTopology
+	}
+}
+
+func (ps *ProbingScheduler) stopWorkerForEndpoint(endpoint topology.ProbeableEndpoint) {
+	workerChan, ok := ps.workerControlChans[endpoint]
+	if ok {
+		level.Info(ps.logger).Log("msg", fmt.Sprintf("Stopping probing on %s", endpoint.GetName()))
+		workerChan <- false // Terminate workers
+		delete(ps.workerControlChans, endpoint)
+	}
+}
+
+func (ps *ProbingScheduler) startNewWorker(endpoint topology.ProbeableEndpoint, checks []Check) {
+	wc := make(chan bool)
+	w := ProberWorker{logger: log.With(ps.logger, "endpoint_name", endpoint.GetName()),
+		endpoint: endpoint, checks: checks,
+		controlChan: wc, refreshInterval: 30 * time.Second}
+	ps.workerControlChans[endpoint] = wc
+	go w.StartProbing()
+}
+
+type ProberWorker struct {
+	logger          log.Logger
+	endpoint        topology.ProbeableEndpoint
+	refreshInterval time.Duration
+	checks          []Check
+	controlChan     chan bool
+}
+
+func (pw *ProberWorker) findWork(lastChecks []time.Time) {
+	for {
+		check_performed := false
+		for i, check := range pw.checks {
+			if lastChecks[i].Add(check.Interval).Before(time.Now()) {
+				level.Debug(pw.logger).Log("msg", fmt.Sprintf("Performing check %s", check.Name))
+				err := check.CheckFn(pw.endpoint)
+				if err != nil {
+					CheckFailureTotal.WithLabelValues(check.Name, pw.endpoint.GetName(), check.Name)
+					level.Error(pw.logger).Log("msg", "Error while probing", "err", err)
+				} else {
+					CheckSuccessTotal.WithLabelValues(check.Name, pw.endpoint.GetName(), check.Name)
+				}
+				lastChecks[i] = time.Now()
+				check_performed = true
+			}
+		}
+		// If we have performed a check it means we lost some time
+		// We should check for available work before going to sleep again
+		if !check_performed {
+			break
+		}
+	}
+}
+
+func (pw *ProberWorker) StartProbing() {
+	level.Info(pw.logger).Log("msg", "starting probing")
+
+	if len(pw.checks) < 1 {
+		level.Error(pw.logger).Log("msg", "Probe not started no checks registered")
+		return
+	}
+
+	lastChecks := make([]time.Time, len(pw.checks))
+	shortestInterval := pw.checks[0].Interval
+
+	err := pw.endpoint.Connect()
+	if err != nil {
+		level.Error(pw.logger).Log("msg", "Probe failure during connection", "err", err)
+		return
+	}
+
+	for i, check := range pw.checks {
+		lastChecks[i] = time.Now()
+		check.PrepareFn(pw.endpoint)
+		if shortestInterval > check.Interval {
```

Comment:
plz add a comment here to explain why we look for shortestInterval
-> shortestInterval will be used as ticker for all checks. findWork will then checks if we wait enough time before to run each checks. Means that we can have an approximate on duration between checks